demoapp/testimony_summary.py

```diff
@@ -0,0 +1,204 @@
+
+open_ai_key, bill_id = sys.argv[1:]
+os.environ['OPENAI_API_KEY'] = open_ai_key 
+MODEL_NAME = 'gpt-4-1106-preview'
+MAX_TOKENS = 4097
+
+mgl_df = pd.read_csv("./12_bills_with_mgl_more_sections.csv")
+summaries = pd.read_csv('./generated_bills.csv')
+testimonies = pd.read_csv('./maple_testimonies.csv')
+
+text_splitter = CharacterTextSplitter.from_tiktoken_encoder(model_name = MODEL_NAME)
+
+bills_to_select = {
+    '#H3121': 'An Act relative to the open meeting law',
+    '#S2064': 'An Act extending the public records law to the Governor and the Legislature',
+    '#H711': 'An Act providing a local option for ranked choice voting in municipal elections',
+    '#S1979': 'An Act establishing a jail and prison construction moratorium',
+    '#H489': 'An Act providing affordable and accessible high-quality early education and care to promote child development and well-being and support the economy in the Commonwealth',
+    '#S2014': 'An Act relative to collective bargaining rights for legislative employees',
+    '#S301': 'An Act providing affordable and accessible high quality early education and care to promote child development and well-being and support the economy in the Commonwealth',
+    '#H3069': 'An Act relative to collective bargaining rights for legislative employees',
+    '#S433': 'An Act providing a local option for ranked choice voting in municipal elections',
+    '#H400': 'An Act relative to vehicle recalls',
+    '#H538': 'An Act to Improve access, opportunity, and capacity in Massachusetts vocational-technical education',
+    '#S257': 'An Act to end discriminatory outcomes in vocational school admissions'
+}
+
+def find_bills(bill_number, df = mgl_df):
+    """input:
+    args: bill_number: (str), Use the number of the bill to find its title and content
+    """
+    bill = df[df['BillNumber'] == bill_number]['DocumentText']
+
+    try:
+         # Locate the index of the bill
+        idx = bill.index.tolist()[0]
+        # Locate the content and bill title of bill based on idx
+        content = df['DocumentText'].iloc[idx]
+        bill_title = df['Title'].iloc[idx]
+        bill_number = df['BillNumber'].iloc[idx]
+        # laws
+        # law = df['combined_MGL'].iloc[idx]
+
+        return content, bill_title, bill_number
+    
+    except Exception as e:
+        content = "blank"
+        print(e)
+
+def generate_categories(text):
+
+    category_prompt = """According to this list of category {category}.
+
+        classify this bill {context} into a closest relevant category.
+
+        Do not output a category outside from the list
+    """
+
+    prompt = PromptTemplate(template=category_prompt, input_variables=["context", "category"])
+    llm = LLMChain(llm = ChatOpenAI(openai_api_key=open_ai_key, temperature=0, model=MODEL_NAME), prompt = prompt)
+        
+    response = llm.predict(context = text, category = category_for_bill) # grab from tagging.py
+    return response
+
+def cache(func):
```

Comment:
Would it be better to use the langchain built in SQLite cache?

https://python.langchain.com/docs/modules/model_io/llms/llm_caching